README.md
# Single Channel LoRaWAN Gateway

Version 6.2.8, 
Data: July 11, 2021  
Author: M. Westenberg (mw12554@hotmail.com)  
Copyright: M. Westenberg (mw12554@hotmail.com)  

All rights reserved. This program and the accompanying materials are made available under the terms 
of the MIT License which accompanies this distribution, and is available at
https://opensource.org/licenses/mit-license.php  
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Maintained by Maarten Westenberg (mw12554@hotmail.com) 

![logo](https://github.com/things4u/ESP-1ch-Gateway/blob/master/images/illustration-Maarten-In.jpg?raw=true "aap")
 
# Description

First of all: PLEASE READ THIS FILE AND [Documentation](http://things4u.github.io/Projects/SingleChannelGateway) 
it should contain most of the information you need to get going.
Unfortunately I do not have the time to follow up on all emails, and as most information including pin-outs 
etc. etc. are contained on these pages I hope you have the time to read them and post any remaining questions.

I do have more than 10 Wemos D1 mini boards running, some I built myself, 
some 10+ on Hallard, 3 on ComResult and 4 ESP32 boards. They ALL work without problems
on this code.
I did find however that good soldering joints and wiring makes all the difference,
so if you get resets/errors you cannot explain, please have a second look at your wiring.

This repository contains a proof-of-concept implementation of a single channel LoRaWAN gateway for the ESP8266. 
Starting version 5.2 also the ESP32 of TTGO (and others) is supported.
The software implements a standard LoRa gateway with the following exceptions and changes:

-  This LoRa gateway is not a full gateway but it implements just a one-channel/one frequency gateway. 
The minimum amount of frequencies supported by a full gateway is 3, most support 9 or more frequencies.
This software started as a proof-of-concept to prove that a single low-cost RRFM95 chip which was present 
in almost every LoRa node in Europe could be used as a cheap alternative to the far more expensive full 
gateways that were making use of the SX1301 chip.

- As the software of this gateway will often be used during the development phase of a project or in 
demo situations, the software is flexible and can be easily configured according to environment or 
customer requirements. There are two ways of interacting with the software: 
1. Modifying the configGway.h file at compile time allows the administrator to set almost all parameters. 
2. Using the webinterface (http://<gateway_IP>) will allow administrators to set and reset several of the 
parameters at runtime.

Full documentation of the Single Channel Gateway is found at things4u.github.io, please look at the Hardware Guide 
under the Gateway chapter.

# PlatformIO or ArduinoIDE

The source works on both environments, both the classic Arduino IDE and on PlatformIO. 
Unfortunately there are small differences between these two envrironments. 
At this moment the src directory contains the PlatformIO source, and therefore we will decribe how to connect to Arduino IDE.
The applies to the libraries.

## PlatformIO
When in PlatformIO, choose <File> and then <Add folder to Workspace...> and select your new LoRa-1ch-ESP-Gateway 
top directory.
Then just open the ESP-sc-gway.ino file at src directory and build or upload

## Arduino IDE

Create a place on you filesystem to work on the files. In this directory create the source directory "ESP-sc-gway" 
and the libraries directory "libraries". When unpacking the source at github: 
Copy the content of the "src" directory to the Aruino IDE "ESP-sc-gway" directory and copy the contents 
of the "lib" directory to the Arduino IDE "libraries" directory;

## testing

The single channel gateway has been tested on a gateway with the Wemos D1 Mini, 
using a HopeRF RFM95W transceiver.  Tests were done on 868 version of LoRa and some 
testing on 433 MHz.
The LoRa nodes tested against this gateway are:

- TeensyLC with HopeRF RFM95 radio
- Arduino Pro-Mini (default Armega328 model, 8MHz 3.3V and 16MHz 3.3V)
- ESP8266 based nodes with RFM95 transceivers.

The code has been tested on at least 8 separate gateway boards both based on the Hallard and the Comresult boards. 
I'm still working on the ESP32 pin-out and functions (expected soon).

# Getting Started

It is recommended to compile and start the single channel gateway with as little modificatons as possible. 
This means that you should use the default settings in the 2 configuration files as much as possible and 
only change the SSID/Password for your WiFi setup and the pins of your ESP8266 that you use in loraModem.h.
This section describes the minimum of configuration necessary to get a working gateway which than can be 
configured further using the webpage.

1. Unpack the source code including the libraries in a separate folder.
2. Connect the gateway to a serial port of your computer, and configure that port in the IDE. 
Switch on the Serial Monitor for the gateway. As the Wemos chip does not contain any code, 
you will probably see nothing on the Serial Monitor.
3. If necessary, modify the _loraModem.h file and change the "struct pins" area and configure 
either for a traditional (=Comresult) PCB or configure for a Hallard PCB where the dio0, 
dio1 and dio2 pins are shared. You HAVE to check this section. In the configGway.h file the 
most used pin-outs are documented so it might be that you use a standard pin-out
4. Edit the configNode.h file and adapt the "wpas" structure. Make sure that the first line of 
this structure remains empty and put the SSID and Password of your router on the second 
line of the array.
5. In the preferences part of the IDE, set the location of your sketch to the place where 
you put the sketch on your computer. This will make sure that for example the required 
libraries that are shipped with this sketch in the libraries folder can be found by the 
compiler.
6. If not yet done: Load the support for ESP8266 in your IDE. <Tools><Board><Board Manager...>
7. Load the other necessary libraries that are not shipped with this sketch in your IDE. 
Goto <Sketch><Include Library><Manage Libraries...> in the IDE to do so. Most of the include files 
can be loaded through this library section. Some cannot and are shipped with the Gateway.


- LoRaCode (Version 1.0.0, see library shipped)
- gBase64 (changed name from Adam Rudd's Base64 version)
- TinyGPS++ (Version 1.0.0)

Through Library Manager:

- ArduinoJson (version 6.12.0)
- Heltec ESP32 Dev-Boards (Version 1.0.6)
- Heltec ESL8266 Dev-Boards (Version 1.0.2)
- SPI (Version 1.0.0)
- SPIFFS (Version 1.0.0)
- Streaming (Version 5.0.0)
- Ticker (Version 1.1.0)
- Time (Version 1.5.0)
- Update (Version 1.0.0)
- Webserver (Version 1.0.0)
- WiFiClientSecure (Version 1.0.0)
- WifiEsp (Version 2.2.2)
- Wire (Version 1.0.1)
- ESP_WifiManager (Version 1.0.2 by Khoi Hoang)

8. Compile the code and download the executable over USB to the gateway. If all is right, you should
see the gateway starting up on the Serial Monitor.
9. Note the IP address that the device receives from your router. Use that IP address in a browser on 
your computer to connect to the gateway with the browser.

Now your gateway should be running. Use the webpage to set "debug" to 1 and you should be able to see packages
coming in on the Serial monitor.


# Configuration

There are two ways of changing the configuration of the single channel gateway:

1. Changing the configGway.h and the configNode.h file at compile-time
2. Run the http://<gateway-IP> web interface to change settings at run time.

Where you have a choice, option 2 is far more friendly and useful.

## Editing the configGway.h file

The configGway.h file contains the user configurable gateway settings. 
All have their definitions set through #define statements. 
In general, setting a #define to 1 will enable the function and setting it to 0 will disable it. 

Also, some settings can be initialised by setting their value with a #define but can be changed at runtime in the web interface.
For some settings, disabling the function with a #define will remove the function from the webserver as well.

NOTE regarding memory usage: The ESP8266 has an enormous amount of memory available for program space and SPIFFS filesystem. 
However the memory available for heap and variables is limited to about 80K bytes (For the ESP-32 this is higher). 
The user is advised to turn off functions not used in order to save on memory usage. 
If the heap drops below 18 KBytes some functions may not behave as expected (in extreme case the program may crash).

## Editing the configNode.h file

The configNode.h file is used to set teh WiFi access point and the structure of known 
sensors to the 1-channel gateway.
Setting the known WiFi access points (SSID and password) must be done at compile time.

When the gateway is not just used as a gateway but also for debugging purposes, the used can specify not only
the name of the sensor node but also decrypt for certain nodes the message.

### Setting USB

The user can determine whether or not the USB console is used for output messages.
When setting _DUSB to 0 all output by Serial is disabled 
(actually the Serial statements are not included in the code).

 \#define _DUSB 1

### Selecting Class mode of operation

Define the class of operation that is supported by the gateway. 
Class A is supported and contains the basic operation for battery sensors. 

Class B contains the beacon/battery mode of operation. The Gateway will send ou a beacon to the 
connected sensors which enables them to synchronize downlink messagaging.

Class C (Continuous) mode of operation contains support for devices that are probably NOT battery 
operated and will listen to the network at all times. As a result the latency of these devices 
is also shorter than for class A devices.
Class C devices are not dependent on battery power and will extend the receive windows
until the next transmission window. In fact, only transmissions will make the device abort 
listening as long as this tranmission lasts. 
Class C devices cannot do Class B operation.

 \#define _CLASS "A"
 
 All devices will start as class A devices, and may decide to "upgrade" to class B or C.
Also the gateway may or may not support Class B, which is a superset of class A.
 NOTE: Only class A is supported
 
### Selecting you standard pin-out

We support five pin-out configurations out-of-the-box, see below.
If you use one of these, just set the parameter to the right value.
If your pin definitions are different, update the loraModem.h and oLED.h file to reflect these settings.
	1: HALLARD
	2: COMRESULT pin out
	3: ESP32/Wemos based board
	4: ESP32/TTGO based ESP32 boarda
	5: ESP32/Heltec Wifi LoRA 32(V2)

 \#define _PIN_OUT 1


### Forcing a SPIFF format at startup

The following parameter shoudl be set to 0 under normal circumstances.
It does allow the system to foce formatting of the SPIFFS filesystem.

 \#define SPIFF_FORMAT 0  
 
### Setting Spreading Factor

Set the _SPREADING factor to the desired SF7, SF8 - SF12 value. 
Please note that this value is closely related to the value used for _CAD. 
If _CAD is enabled, the value of _SPREADING is not used by the gateway as it has all sreading factors enabled.

 \#define _SPREADING SF9
 
Please note that the default frequency used is 868.1 MHz which can be changed in the loraModem.h file. 
The user is advised NOT to change this etting and only use the default 868.1 MHz frequency.


### Channel Activity Detection

Channel Activity Detection (CAD) is a function of the LoRa RFM95 chip to detect incoming messages (activity). 
These incoming messages might arrive on any of the well know spreading factors SF7-SF12. 
By enabling CAD, the gateway can receive messages of any of the spreading factors.

Actually it is used in normal operation to tell the receiver that another signal is using the 
channel already.

The CAD functionality comes at a (little) price: The chip will not be able to receive very weak signals as 
the CAD function will use the RSSI register setting of the chip to determine whether or not it received a 
signal (or just noise). As a result, very weak signals are not received which means that the range of the 
gateway will be reduced in CAD mode.

 \#define _CAD 1

 
### Over the Air Updates (OTA)

As from version 4.0.6 the gateway allows over the air updating if the setting A_OTA is on. 
The over the air software requires once setting of the 4.0.6 version over USB to the gateway,
after which the software is (default) enabled for use.

The first release only supports OTA function using the IDE which in practice means the IDE has to 
be on the same network segment as the gateway.

Note: You have to use Bonjour software (Apple) on your network somewhere. A version is available
for most platforms (shipped with iTunes for windows for example). The Bonjour software enables the
gateway to use mDNS to resolve the gateway ID set by OTA after which download ports show up in the IDE.

Todo: The OTA software has not (yet) been tested in conjuction with the WiFiManager software.

 \#define A_OTA 1  



### Enable Webserver

This setting enables the webserver. Although the webserver itself takes a lot of memory, it greatly helps 
to configure the gatewayat run-time and inspects its behaviour. It also provides statistics of last messages received.
The A_REFRESH parameter defines whether the webserver should renew every X seconds.

 \#define A_SERVER 1				// Define local WebServer only if this define is set  
 \#define A_REFRESH 1				// is the webserver enabled to refresh yes/no?  (yes is OK)
 \#define A_SERVERPORT 80			// local webserver port  
 \#define A_MAXBUFSIZE 192			// Must be larger than 128, but small enough to work  

 The A_REFRESH parameter defines whether or not we can set the refresh yes/no setting in the webbrowser. 
 The setting in the webbrowser is normally put on "no" as a default, but we can leave the define on
 "1" to enabled that setting in the webbrowser.
 
### Strict LoRa behaviour

In order to have the gateway send downlink messages on the pre-set spreading factor and 
on the default frequency, you have to set the _STRICT_1CH parameter to 1. Note that when 
it is not set to 1, the gateway will respond to downlink requests with the frequency 
and spreading factor set by the backend server. And at the moment TTN responds to downlink 
messages for SF9-SF12 in the RX2 timeslot and with frequency 869.525MHz and on SF12 
(according to the LoRa standard when sending in the RX2 timeslot). 

 \#define _STRICT_1CH 0

You are advised not to change the default setting of this parameter.

### Enable OLED panel

By setting the OLED you configure the system to work with OLED panels over I2C.
Some panels work by both SPI and I2C where I2c is solwer. However, since SPI is use for RFM95 transceiver 
communication you are stronly discouvared using one of these as they will not work with this software.
Instead choose a OLED solution that works over I2C.

 \#define OLED 1

The following values are defined for OLED:
1. 0.9 inch OLED screen for I2C bus
2. 1.1 inch OLED screen for I2C bus

### Define to gather statistics

When this is defined (==1) we will gather the statistics of every message and output
it to the SPIFFS filesystem. We make sure that we use a number of files with each a fixed number of records
for statistics. The REC number tells us how many records are allowed in each statistics file.
As soon as the REC number is higher than the number of records allowed, we open a new file.

CHANGELOG.md
# Single Channel LoRaWAN Gateway

Version 6.2.8, July 11, 2021 
Author: M. Westenberg (mw12554@hotmail.com)  
Copyright: M. Westenberg (mw12554@hotmail.com)  

All rights reserved. This program and the accompanying materials are made available under the terms 
of the MIT License which accompanies this distribution, and is available at
https://opensource.org/licenses/mit-license.php  
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Maintained by Maarten Westenberg (mw12554@hotmail.com)



# Release Notes

Release 6.2.8 (July 11, 2021)
- Change buttons on top of web screen to work with Firefox without displaying a menu
- Added South Korea country to loraModem.h
- Removed superfluous comments in all files
- Removed lots of superfluous mPrint and other debug lines
- Show register values (menu top of web page)
- repair syntax errors
- Add configurable parameters to platformIO and other IDE in configGateway.h
- Make more and detailed documentation available
- In configGway.h change default server to TTN v3. 
	#	define _TTNSERVER "eu1.cloud.thethings.network"
- Change the Documentation option for the GUI such that clicking it will open a separate tab/window.

Features  release 6.2.7 (October 30, 2020)
- Further repair timing
- Make interrupt handling simplier
- improve website of gateway, making data dynamic

Features release 6.2.6 (September 08, 2020)
- Better timing for downstream
- Other display of statistics

Features release 6.2.5 (April 30, 2020)
- Repaired SF and BW for upstream
- Rewrote Monitoring to output only the most significant messages for debug==1
- Make definitions of LASTSEEN and PACKAGE STATISTICS dynamic. Allowing GUI resizing. 
Look in expert mode to change these settings
- Rewrote the addLog() function
- Repaired typos
- Removed unncessary serial and MONITOR prints

Features release 6.2.4 (April 25, 2020)
- Changes the date layout used in the output to be more standard: 3 characters weekday, months and day 2 chars
- Changed the configGway.h file a lot ro define default values for parameters while at the 
	same time allowing changing through PlatformIO
- Updated the documentation itself and updated link in the software
- For Monitor output use the same stringTime() function
- Added the _PROFILER definition to read the timing (PlatformIO only)
- Adapt the documentation and start writing hardware guide (for compile options)
- Changed the delay for TX (Downlink messages) -. Removed to 0 for _STIRCT_1CH
- Index of iSeen messages was in Hex -> Changed to decimal
- Added a timing correction for sending message of aound 700000 uSec (0.7 Sec) to make downlink work
- Changed naming of tmst (TimeStamp) functions
- Removed function printlog(), as it was not used
- Moved the SPIFFS file operations to a timer function in loop() as it consumer far too much time. 
As a result regular timestamp is reduced.
- Removed the logging functions for the same reasons. User can choose to put those on again.
- Added Udp.flush() when NTP messages arrive during operations 
- Changed layout of webpage. Changed sequence to first OFF and then ON
- Changed website to include loraWait statistics

Features release 6.2.3 (February 23, 2020)
- Lots of bugs and documentation fixes
- Added customizable #define statements through platformio.ini file (Read!!!)
- Changed the WiFiManager code to better support both architectures: ESP8266 and ESP32

Features release 6.2.1 (February 2, 2020)
- PlatformIO support

Features release 6.2.0 (January 29, 2020)
- Indicate WPA and WIFI mode on OLED display
- Correct sensor totals in Webserver overview.
- Change delays for GPS and other internal sensors, and repair internal 
  sensors to make sure we do not miss sensor messages
- Changed the MONITOR part to a circular buffer to save time when filling buffer
- Statistics log redefined
- Removed variables never used (PackaIO)


Features release 6.1.8 (January 21, 2020)
- Repair wifimanager compile for ESP8266, and make it work for ESP8266
- Substitues _DUSB by _MONITOR in a lot of cases
- Repair c_str() bugs
- Make getNtpTime more predictable
- Several syntax error optimizations
- InLine documentation updated and errors corrected
- Repair Format code
- Repair WLAN Connect code, remove code to make connecting lean.
- Add a lot of output to MONITOR screen
- Repair STAT_LOG so enable running without log output
- Add '_' to front of defined names

Features release 6.1.7 (December 27, 2019)
- Repair bugs for DNS
- Enable internal T-Beam sensors to output values in "raw" format. 
See configNode.h and _sensors.ino

Features release 6.1.5 (December 20, 2019)
- Bug fix for "#define _DUSB 0"
- Added number of times a node has been called to the history records listSeen
- Fix LastSeen issues.
- Added 3 Indian frequencies as published by TTN
- Changes configGway.h to contain better info
- Added a message console on the GUI page (_define _MAXMONITOR 1)
- Changed the DNS code to allow .local domains with MDNS
- Added buttons for Expert mode, Monitor and Last Seen.
- Added code for "REBOOT" to be used from the API only *http://your_IP/REBOOT '
- Added the IP to the www server header (visible in browser).
- Removed version 5 of pi-out as it is the same as version 4.
- Added "monitor" functions. This allows to write to monitor screen on browser or use 
USB for desbugging messages.

Features release 6.1.4 (November 25, 2019)
- Compacting Code and Solve Errors
- Look at _DUSB define and add to Serial.print directive where not found
- Renewed the GPS functions, changed "Serial1" to "sGps" to avoid double definitions.
- Downloaded TinyGPS++ library. All working on my T-Beam again.

Features release 6.1.3 (November 20, 2019)
- Made changes to _TRUSTED_NODES in _wwwServer.ino to make sure only named nodes are 
displayed when the value of trusted nodes in "Gateways Settings" has value 2. 

src/configGway.h
// 1-channel LoRa Gateway for ESP32 and ESP8266
// Copyright (c) Maarten Westenberg 2016-2021 

#define VERSION "V.6.2.8.EU868; PlatformIO 210712 a; GIT"

//
// Based on work done by Thomas Telkamp for Raspberry PI 1ch gateway and many others.
//
// All rights reserved. This program and the accompanying materials
// are made available under the terms of the MIT License
// which accompanies this distribution, and is available at
// https://opensource.org/licenses/mit-license.php
//
// NO WARRANTY OF ANY KIND IS PROVIDED
//
// Author: Maarten Westenberg (mw12554@hotmail.com)
//
// This file contains a number of compile-time settings that can be set on (=1) or off (=0)
// The disadvantage of compile time is minor compared to the memory gain of not having
// too much code compiled and loaded on your ESP device.
//
// NOTE: 
// If version is for ESP32 Heltec board, compile with ESP32 setting and board 
// "ESP32 Dev Module" or "Heltec WiFi Lora 32"
// 
// For ESP8266 Wemos: compile with "Wemos R1 D1" and choose
// the right _PIN_OUT below. Selecting Oled while that is not connected does not 
// really matter (so you can leave that in).
//
// The source has been optimized for PlatformIO usage. That means some variables
// can be defined in the .platformio.ini file. Please look at the example file to see 
// how to set these #defines
//
// ========================================================================================


// Define whether we should do a formatting of SPIFFS when starting the gateway
// This is usually a good idea if the webserver is interrupted halfway a writing
// operation. Also to be used when software is upgraded
// Normally, value 0 is a good default and should not be changed.
#if !defined _SPIFFS_FORMAT
#	define _SPIFFS_FORMAT 0
#endif


// Allows configuration through WifiManager AP setup. Must be 0 or 1	
#if !defined _WIFIMANAGER				
#	define _WIFIMANAGER 0
#endif


// Debug message will be put on Serial is this one is set.
// If set to 0, no printing to USB devices is done.
// Set to 1 it will print all user level messages (with correct debug set)
// If set to 2 it will also print interrupt messages (not recommended)
#if !defined _DUSB
#	define _DUSB 1
#endif


// Define the monitor screen. When it is greater than 0 then logging is displayed in
// the special screen at the GUI.
// If _DUSB is also set to 1 then most messages will also be copied to USB devices.
#if !defined _MONITOR
#	define _MONITOR 1
#endif


// Gather statistics on sensor and Wifi status
// 0= No statistics
// 1= Keep track of messages statistics, number determined by _MAXSTAT
// 2= Option 1 + Keep track of messages received PER each SF (default)
// 3= See Option 2, but with extra channel info (Not used when Hopping is not selected)
#if !defined _STATISTICS
#	define _STATISTICS 3
#endif

// Define the frequency band the gateway will listen on. Valid options are
// EU863_870	Europe 
// US902_928	North America
// AU925_928	Australia
// CN470_510	China
// IN865_867	India
// CN779-787	(Not Used)
// EU433		Europe
// KR920    	South Korea
// AS923		(Not Used)
// You can find the definitions in "loraModem.h" and frequencies in
// See https://www.thethingsnetwork.org/docs/lorawan/frequency-plans.html
#define EU863_870 1
 

// Define the CLASS mode of the gateway
// A: Baseline Class
// B: Beacon/Battery Class
// C: Continuous Listen Class
#define _CLASS "A"


// Define whether to use the old Semtech gateway API, which is still supported by TTN,
// but is more lightweight than the new TTN tcp based protocol.
// NOTE: Only one of the two should be defined! TTN Router project has stopped
//
#define _UDPROUTER 1
//#define _TTNROUTER 1


#if !defined _CHANNEL
#	define _CHANNEL 0
#endif

// The spreading factor is the most important parameter to set for a single channel
// gateway. It specifies the speed/datarate in which the gateway and node communicate.
// As the name says, in principle the single channel gateway listens to one channel/frequency
// and to one spreading factor only.
// This parameters contains the default value of SF, the actual version can be set with
// the webserver and it will be stored in SPIFF
// NOTE: The frequency is set in the loraModem.h file and is default 868100000 Hz.
#if !defined _SPREADING
#	define _SPREADING SF9
#endif

// Channel Activity Detection
// This function will scan for valid LoRa headers and determine the Spreading 
// factor accordingly. If set to 1 we will use this function which means the 
// 1-channel gateway will become even more versatile. If set to 0 we will use the
// continuous listen mode.
// Using this function means that we HAVE to use more dio pins on the RFM95/sx1276
// device and also connect enable dio1 to detect this state. 
#define _CAD 1


// CRCCHECK
// Defines whether we should check on the CRC of RXDONE messages (see stateMachine.ino)
// This should prevent us from getting a lot os stranges messgages of unknown nodes.
// Note: DIO3 must be connected for this to work (Heltec and later Wemos gateways). 
#define _CRCCHECK 1


// Definitions for the admin webserver.
// _SERVER determines whether or not the admin webpage is included in the sketch.
// Normally, leave it in!
#define _SERVER 1				// Define local WebServer only if this define is set
#define _REFRESH 1				// Allow the webserver refresh or not?
#define _SERVERPORT 80			// Local webserver port (normally 80)
#define _MAXBUFSIZE 192			// Must be larger than 128, but small enough to work


// Definitions for over the air updates. At the moment we support OTA with IDE
// Make sure that tou have installed Python version 2.7 and have Bonjour in your network.
// Bonjour is included in iTunes (which is free) and OTA is recommended to install 
// the firmware on your router witout having to be really close to the gateway and 
// connect with USB.
#if !defined _OTA
#	define _OTA 1
#endif


// We support a few pin-out configurations out-of-the-box: HALLARD, COMPRESULT and 
// Heltec/TTGO ESP32.
// If you use one of these, just set the parameter to the right value.
// If your pin definitions are different, update the loraModem.h file to reflect the
// hardware settings.
//	1: HALLARD
//	2: COMRESULT pin out
//	3: ESP32, Wemos pin out (Not used)
//	4: ESP32, Heltec and TTGO pin out (should work for Heltec, 433 and Oled too).
//	5: Other, define your own in loraModem.h (does not include GPS Code)
#if !defined _PIN_OUT
#	define _PIN_OUT 1
#endif


// Single channel gateways if they behave strict should only use one frequency 
// channel and one, or in case _CAD all, spreading factors. 
// The TTN backend replies on RX1 timeslot for spreading factors SF9-SF12. 
// If the 1ch gateway is working in and for nodes that ONLY transmit and receive on the set
// and agreed frequency and spreading factor. make sure to set STRICT to 1.
// In this case, the frequency and spreading factor for downlink messages is adapted by this
// gateway
// NOTE: If your node has only one frequency enabled and one SF, you must set this to 1
//		in order to receive downlink messages. This is the default mode.
// NOTE: In all other cases, value 0 works for most gateways with CAD enabled
#if !defined _STRICT_1CH
#	define _STRICT_1CH 1
#endif


// Extra Microseconds delay added by the Up receiver from the sensor to the server.
// As the tmst is also corrected, this will add to the downlink messages also.
// The server will use this value to compute the receive window.
#if !defined _RXDELAY1
//#	define _RXDELAY1 1000
#	define _RXDELAY1 0
#endif


//
// Also, normally the server will respond with SF12 in the RX2 timeslot.
// For TTN, the RX2 timeslot is SF9, so we should use that one for TTN
#if !defined _RX2_SF
#	define _RX2_SF 9
#endif


// This section defines whether we use the gateway as a repeater
// For his, we use another output channel as the channel (default==0) we are 
// receiving the messages on.
#if !defined _REPEATER
#	define _REPEATER 0
#endif


// Define if Oled Display is connected to I2C bus. Note that defining an Oled display does not
// impact performance negatively, certainly if no Oled is connected. Wrong Oled will not show
// sensible results on the Oled display
// _OLED==0;	No Oled display connected
// _OLED==1;	0.9" Oled Screen based on SSD1306
// _OLED==2;	1.3" Oled screens for Wemos, 128x64 SH1106
#if !defined _OLED
#	define _OLED 1
#endif


// Define whether we want to manage the gateway over UDP (next to management 
// thru webinterface).
// This will allow us to send messages over the UDP connection to manage the gateway 
// and its parameters. Sometimes the gateway is not accesible from remote, 
// in this case we would allow it to use the SERVER UDP connection to receive 
// messages as well.
// NOTE: Be aware that these messages are NOT LoRa and NOT LoRa Gateway spec compliant.
//	However that should not interfere with regular gateway operation but instead offer 
//	functions to set/reset certain parameters from remote.
#define _GATEWAYMGT 0


// Do extensive logging to file(s)
// Use the ESP8266 SPIFS filesystem to do extensive logging.
// We must take care that the filesystem never(!) is full, and for that purpose we
// rather have new records/line of statistics than very old.
// Of course we must store enough records to make the filesystem work
// NOTE:
// Please do NOT USE file logging if you are concerned about downstream timing.
//	The status logging taken about 6 ms (6000 uSec) for each write whereas
//	normally this is less than 100 uSecs.
#if !defined _STAT_LOG
#	define _STAT_LOG 0
#endif


// Set the Server Settings (IMPORTANT)
#define _LOCUDPPORT 1700					// UDP port of gateway! Often 1700 or 1701 is used for upstream comms


// This defines whether or not we would use the gateway as 
// as sort of backend decoding system for local sensors which decodes (such as TTGO T-Beam)
// 0: Do not use _LOCALSERVER
// 1: _LOCALSERVER is used for received messages
// 2: Also transmittes messages are encoded
#if !defined _LOCALSERVER
#	define _LOCALSERVER 1					// See server definitions for decodes
#endif


// ntp
// Please add daylight saving time to NTP_TIMEZONES when desired
#define NTP_TIMESERVER "nl.pool.ntp.org"	// Country and region specific
#define NTP_TIMEZONES	2					// How far is our Timezone from UTC (excl daylight saving/summer time)
#define SECS_IN_HOUR	3600
#define _NTP_INTR 0							// Do NTP processing with interrupts or in loop();


// lora sensor code definitions
// Defines whether the gateway will also report sensor/status value on MQTT 
// (such as battery and GPS)
// after all, a gateway can be a node to the system as well. Some sensors like GPS can be
// sent to the backend as a parameter, some (like humidity for example) can only be sent
// as a regular sensor value.
// Set its LoRa address and key below in this file, See spec. para 4.3.2
// NOTE: The node is switched off by default. Switch it on in the GUI
#if !defined _GATEWAYNODE
#	define _GATEWAYNODE 0
#endif


// We can put the gateway in such a mode that it will (only) recognize
// nodes that are put in a list of trusted nodes.
// Values:
// 0: Do not use names for trusted Nodes
// 1: Use the nodes as a translation table for hex codes to names (in TLN)
// 2: Same as 1, but is nodes NOT in the nodes list below they are NOT shown
// NOTE: We probably will make this list dynamic!
#define _TRUSTED_NODES 1


// ========================================================================
// DO NOT CHANGE BELOW THIS LINE
// Probably do not change items below this line, only if lists or 
// configurations on configNode.h are not large enough for example.
// ========================================================================


// Name of he configfile in SPIFFs	filesystem
// In this file we store the configuration and other relevant info that should
// survive a reboot of the gateway		
#define _CONFIGFILE "/gwayConfig.txt"


// Maximum number of Message History statistics records gathered. 20 is a good maximum 
// (memory intensive). For ESP32 maybe 30 could be used as well
#if !defined _MAXSTAT
#	define _MAXSTAT 20
#endif


// We will log a list of LoRa nodes that was forwarded using this gateway.
// For eacht node we record:
//	- node Number, or known node name
//	- Last seen 'seconds since 1/1/1970'
//	- SF seen (8-bit integer with SF per bit)
// The initial version _NUMMAX stores max this many nodes, please make
// _MAXSEEN==0 when not used
#if !defined _MAXSEEN
#	define _MAXSEEN 20
#endif
#define _SEENFILE "/gwaySeen.txt"


// Define the maximum amount of items we monitor on the screen
#if !defined _MAXMONITOR
#	define _MAXMONITOR 20
#endif


// Timing
#define _PULL_INTERVAL 16					// PULL_DATA messages to server to get downstream in seconds
#define _STAT_INTERVAL 120					// Send a 'stat' message to server
#define _NTP_INTERVAL 3600					// How often do we want time NTP synchronization
#define _WWW_INTERVAL 60					// Number of seconds before we refresh the WWW page
#define _FILE_INTERVAL 30					// Number of timer (in secs) before writing to files
#define _MSG_INTERVAL 31					// Message timeout timer in seconds
#define _RST_INTERVAL 125					// Reset interval in seconds, total chip reset


// Define the correct radio type that you are using
#define CFG_sx1276_radio		
//#define CFG_sx1272_radio


// Serial Port speed
#define _BAUDRATE 115200						// Works for debug messages to serial momitor


// Will we use Mutex or not?
// +SPI is input for SPI, SPO is output for SPI
#if !defined _MUTEX
#	define _MUTEX 0
#endif //MUTEX

// Read server/gateway traffic
#if !defined _GWAYSCAN
#	define _GWAYSCAN 0						// default, NO _GWAYSCAN
#endif //GWAYSCAN


// Init EXPERT mode
#if !defined _EXPERT
#	define _EXPERT 0						// default, NO _EXPERT
#endif //EXPERT

// MQTT definitions, these settings should be standard for TTN
// and need no changing. When _REPEATER function is selected, we do not
// use the backend function to send message to server over MQTT.
#if _REPEATER==0
// TTN Version 2:
//#	define _TTNSERVER "router.eu.thethings.network"
//
// TTN Version 3:
#	define _TTNSERVER "eu1.cloud.thethings.network"	
#	define _TTNPORT 1700							// Standard port for TTN
#endif
